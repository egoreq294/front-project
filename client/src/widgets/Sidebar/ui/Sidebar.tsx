import cn from 'classnames';
import React, { FC, memo, useState } from 'react';

import { SidebarContent } from './SidebarContent';

import styles from './styles.module.scss';

interface SidebarProps {
  className?: string;
}

export const Sidebar: FC<SidebarProps> = memo(({ className }) => {
  const [collapsed, setCollapsed] = useState<boolean>(false);

  return (
    <aside
      data-testid="sidebar"
      className={cn(
        styles.Sidebar,
        { [styles.Collapsed]: collapsed },
        className,
      )}
    >
      <SidebarContent collapsed={collapsed} setCollapsed={setCollapsed} />
    </aside>
  );
});
