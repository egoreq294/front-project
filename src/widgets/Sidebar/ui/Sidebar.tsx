import cn from 'classnames';
import React, { FC, memo, useState } from 'react';
import { useSelector } from 'react-redux';

import { LanguageSwitcher } from '@features/LanguageSwitcher';
import { ThemeSwitcher } from '@features/ThemeSwitcher';
import { ToggleFeatures } from '@shared/lib/features';
import { AppLogo } from '@shared/ui/deprecated/AppLogo';
import { Button } from '@shared/ui/deprecated/Button';
import { VStack } from '@shared/ui/deprecated/Stack';
import { getSidebarItems } from '../model/selectors/getSidebarItems';
import { SidebarItem } from './SidebarItem/SidebarItem';

import styles from './styles.module.scss';

interface SidebarProps {
  className?: string;
}

export const Sidebar: FC<SidebarProps> = memo(({ className }) => {
  const [collapsed, setCollapsed] = useState<boolean>(false);

  const onToggle = (): void => {
    setCollapsed((prev) => !prev);
  };

  const sidebarItems = useSelector(getSidebarItems);

  return (
    <ToggleFeatures
      feature="redesignedApp"
      on={
        <aside
          data-testid="sidebar"
          className={cn(
            styles.SidebarRedesigned,
            { [styles.Collapsed]: collapsed },
            className,
          )}
        >
          <AppLogo className={styles.AppLogo} />
        </aside>
      }
      off={
        <aside
          data-testid="sidebar"
          className={cn(
            { [styles.Collapsed]: collapsed },
            styles.Sidebar,
            className,
          )}
        >
          <Button
            data-testid="sidebar_toggle"
            className={styles.CollapseButton}
            onClick={onToggle}
            variant="PrimaryInverted"
          >
            {collapsed ? '>' : '<'}
          </Button>
          <VStack
            role="navigation"
            gap="16"
            align={collapsed ? 'center' : 'start'}
            className={styles.Items}
          >
            {sidebarItems.map((item) => (
              <SidebarItem key={item.path} item={item} collapsed={collapsed} />
            ))}
          </VStack>
          <div className={styles.Switchers}>
            <ThemeSwitcher />
            <LanguageSwitcher isShort={collapsed} />
          </div>
        </aside>
      }
    />
  );
});
