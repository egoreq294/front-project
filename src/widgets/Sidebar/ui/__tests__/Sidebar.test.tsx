import { fireEvent, screen } from '@testing-library/react';
import { Sidebar } from '../Sidebar';
import React from 'react';
import { renderWithTranslation } from '@shared/lib/tests/renderWithTranslation/renderWithTranslation';

describe('Sidebar test suite', () => {
  test('should render correctly', () => {
    const { container } = renderWithTranslation(<Sidebar />);

    expect(screen.getByTestId('sidebar')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('should toggle sidebar', () => {
    renderWithTranslation(<Sidebar />);
    const sidebar = screen.getByTestId('sidebar');
    const toggleButton = screen.getByTestId('sidebar_toggle');

    expect(sidebar).toBeInTheDocument();
    expect(toggleButton).toBeInTheDocument();

    fireEvent.click(toggleButton);
    expect(sidebar).toHaveClass('Collapsed');
  });
});
