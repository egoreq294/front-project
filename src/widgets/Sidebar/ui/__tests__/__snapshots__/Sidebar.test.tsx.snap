// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar test suite should render correctly 1`] = `
<div>
  <aside
    class="Sidebar"
    data-testid="sidebar"
  >
    <div
      class="Flex JustifyCenter AlignCenter DirectionRow FullWidth AppLogo AppLogo"
    >
      <div
        class="SmallGradient"
      />
      <div
        class="BigGradient"
      />
      <svg
        fill="currentColor"
        height="40"
        viewBox="0 0 100 40"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4.77 4.235C5.03 3.001 6.26 2 7.513 2h6.812L8.66 28.823H1.848c-1.254 0-2.06-1-1.799-2.235L4.77 4.235ZM27.477 4.235C27.738 3.001 28.967 2 30.22 2h6.812l-5.665 26.823h-6.812c-1.254 0-2.06-1-1.799-2.235l4.721-22.353ZM72.892 4.235C73.152 3.001 74.38 2 75.635 2h6.812l-5.665 26.823h-6.813c-1.254 0-2.059-1-1.798-2.235l4.72-22.353ZM39.303 2h6.812c1.254 0 2.06 1 1.799 2.235l-4.721 22.353c-.26 1.235-1.489 2.235-2.743 2.235h-6.812L39.303 2ZM84.718 2h6.812c1.254 0 2.06 1 1.799 2.235l-4.722 22.353c-.26 1.235-1.488 2.235-2.742 2.235h-6.813L84.718 2ZM50.185 4.235C50.445 3.001 51.673 2 52.927 2h6.813l-5.666 26.823h-6.812c-1.254 0-2.06-1-1.798-2.235l4.72-22.353ZM62.01 2h6.813c1.254 0 2.059 1 1.798 2.235l-7.081 33.53C63.278 38.999 62.05 40 60.796 40h-6.813L62.01 2ZM12.819 19.882h9.083l-1.416 6.706c-.261 1.235-1.49 2.235-2.743 2.235H10.93l1.888-8.94ZM44.52 31.059h9.082L51.714 40h-6.812c-1.255 0-2.06-1-1.799-2.235l1.416-6.706ZM69.174 33.138l-1.15 5.446c-.05.234-.128.298-.366.298h-.523c-.238 0-.29-.064-.24-.298l1.15-5.446c.05-.233.128-.298.366-.298h.523c.238 0 .29.065.24.298ZM70.686 36.812h-.107c-.114 0-.154.032-.177.145l-.344 1.627c-.05.234-.129.298-.366.298h-.524c-.237 0-.289-.064-.24-.298l1.15-5.446c.05-.233.13-.298.367-.298h1.08c1.244 0 1.715.443 1.485 1.53l-.192.911c-.23 1.088-.888 1.53-2.132 1.53Zm.316-2.699-.3 1.426c-.025.113.001.145.116.145h.172c.4 0 .615-.161.702-.572l.12-.572c.087-.41-.059-.572-.46-.572h-.172c-.114 0-.154.032-.178.145ZM74.74 34.991l.85.935c.446.483.508.773.394 1.313l-.03.145c-.215 1.015-.727 1.579-1.914 1.579-1.186 0-1.479-.475-1.205-1.773l.034-.16c.05-.234.129-.299.366-.299h.556c.238 0 .29.065.24.298l-.075.355c-.068.322.036.451.322.451.287 0 .443-.12.505-.41l.032-.154c.048-.226.022-.338-.224-.604l-.8-.862c-.448-.475-.505-.75-.391-1.29l.037-.176c.215-1.015.727-1.58 1.913-1.58 1.187 0 1.48.476 1.206 1.773l-.034.161c-.05.234-.129.298-.366.298h-.557c-.237 0-.289-.064-.24-.298l.075-.354c.068-.323-.035-.451-.322-.451-.286 0-.443.12-.504.41l-.029.137c-.05.234-.024.347.161.556ZM79.532 33.138c.05-.233.128-.298.366-.298h.523c.238 0 .29.065.24.298l-.856 4.053c-.274 1.297-.767 1.772-1.954 1.772-1.186 0-1.479-.475-1.205-1.773l.856-4.052c.05-.233.129-.298.366-.298h.524c.237 0 .289.065.24.298l-.897 4.246c-.068.322.044.451.355.451.302 0 .477-.129.545-.451l.897-4.246ZM82.938 36.417c.003.065.024.08.065.08.04 0 .069-.015.099-.08l1.414-3.367c.069-.17.151-.21.356-.21h.794c.237 0 .289.065.24.298l-1.15 5.446c-.05.234-.13.298-.367.298h-.376c-.237 0-.29-.064-.24-.298l.552-2.61c.015-.072.002-.089-.047-.089-.033 0-.07.017-.09.073l-1.142 2.659c-.082.193-.187.265-.424.265H82.4c-.246 0-.32-.072-.32-.265l-.028-2.66c-.005-.056-.018-.072-.059-.072-.049 0-.069.017-.084.089l-.551 2.61c-.05.234-.128.298-.366.298h-.376c-.238 0-.29-.064-.24-.298l1.15-5.446c.05-.233.129-.298.366-.298h.68c.286 0 .378.065.376.347l-.011 3.23ZM100 2c0 1.105-.89 2-1.987 2a1.993 1.993 0 0 1-1.987-2c0-1.105.89-2 1.987-2S100 .895 100 2Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="Flex JustifyStart AlignStart DirectionColumn Items"
      role="navigation"
    >
      <a
        aria-current="page"
        class="AppLink ActiveLink Link"
        href="/"
      >
        <div
          class="Item"
        >
          <svg
            class="Icon"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.82369 17.9664V13.5316C7.82369 12.9793 8.27141 12.5316 8.82369 12.5316H11.1715C11.7238 12.5316 12.1715 12.9793 12.1715 13.5316V17.9664C12.1715 18.5642 12.6606 19.0533 13.2584 19.0533H16.5193C17.1171 19.0533 17.6062 18.5642 17.6062 17.9664V11.3577C17.6062 10.8054 18.0539 10.3577 18.6062 10.3577H19.454C19.954 10.3577 20.1932 9.73817 19.8127 9.41208L10.7258 1.22737C10.3128 0.857805 9.68237 0.857805 9.26933 1.22737L0.182446 9.41208C-0.187116 9.73817 0.0411431 10.3577 0.541139 10.3577H1.38895C1.94124 10.3577 2.38895 10.8054 2.38895 11.3577V17.9664C2.38895 18.5642 2.87808 19.0533 3.4759 19.0533H6.73674C7.33456 19.0533 7.82369 18.5642 7.82369 17.9664Z"
              fill="currentColor"
            />
          </svg>
          home-link
        </div>
      </a>
      <a
        class="AppLink Link"
        href="/about"
      >
        <div
          class="Item"
        >
          <svg
            class="Icon"
            height="20"
            viewBox="0 0 18 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0.75 2.5C0.75 1.83696 1.01339 1.20107 1.48223 0.732233C1.95107 0.263392 2.58696 0 3.25 0H14.75C15.0783 0 15.4034 0.0646644 15.7067 0.190301C16.01 0.315938 16.2856 0.500087 16.5178 0.732233C16.7499 0.964379 16.9341 1.23998 17.0597 1.54329C17.1853 1.84661 17.25 2.1717 17.25 2.5V16.75C17.25 16.9489 17.171 17.1397 17.0303 17.2803C16.8897 17.421 16.6989 17.5 16.5 17.5H3.03538C2.60163 17.5 2.23618 17.9004 2.54289 18.2071C2.73043 18.3946 2.98478 18.5 3.25 18.5H16.5C16.6989 18.5 16.8897 18.579 17.0303 18.7197C17.171 18.8603 17.25 19.0511 17.25 19.25C17.25 19.4489 17.171 19.6397 17.0303 19.7803C16.8897 19.921 16.6989 20 16.5 20H3.25C2.58696 20 1.95107 19.7366 1.48223 19.2678C1.01339 18.7989 0.75 18.163 0.75 17.5V2.5ZM9 6C9.26522 6 9.51957 5.89464 9.70711 5.70711C9.89464 5.51957 10 5.26522 10 5C10 4.73478 9.89464 4.48043 9.70711 4.29289C9.51957 4.10536 9.26522 4 9 4C8.73478 4 8.48043 4.10536 8.29289 4.29289C8.10536 4.48043 8 4.73478 8 5C8 5.26522 8.10536 5.51957 8.29289 5.70711C8.48043 5.89464 8.73478 6 9 6ZM8.25 7.75V12.75C8.25 12.9489 8.32902 13.1397 8.46967 13.2803C8.61032 13.421 8.80109 13.5 9 13.5C9.19891 13.5 9.38968 13.421 9.53033 13.2803C9.67098 13.1397 9.75 12.9489 9.75 12.75V7.75C9.75 7.55109 9.67098 7.36032 9.53033 7.21967C9.38968 7.07902 9.19891 7 9 7C8.80109 7 8.61032 7.07902 8.46967 7.21967C8.32902 7.36032 8.25 7.55109 8.25 7.75Z"
              fill="currentColor"
            />
          </svg>
          about-link
        </div>
      </a>
    </div>
    <button
      class="IconButton CollapseButton"
      data-testid="sidebar_toggle"
      type="button"
    >
      <svg
        height="20"
        viewBox="0 0 8 13"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.22572 0.862914C6.8407 0.466965 6.20759 0.458106 5.81164 0.843129L0.793132 5.72314C0.576073 5.93421 0.475343 6.21983 0.49206 6.50006C0.475343 6.7803 0.576073 7.06592 0.793132 7.27699L5.81164 12.157C6.20759 12.542 6.8407 12.5332 7.22572 12.1372C7.61074 11.7413 7.60188 11.1082 7.20593 10.7231L2.863 6.50006L7.20593 2.27699C7.60188 1.89197 7.61074 1.25886 7.22572 0.862914Z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="Switchers"
    >
      <button
        class="IconButton"
        type="button"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20C15.523 20 20 15.523 20 10C20 4.477 15.523 0 10 0C4.477 0 0 4.477 0 10C0 15.523 4.477 20 10 20ZM10 18.5V1.5C12.2543 1.5 14.4163 2.39553 16.0104 3.98959C17.6045 5.58365 18.5 7.74566 18.5 10C18.5 12.2543 17.6045 14.4163 16.0104 16.0104C14.4163 17.6045 12.2543 18.5 10 18.5Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="Button Ghost S"
        type="button"
      >
        language
      </button>
    </div>
  </aside>
</div>
`;
